### ISSUE TITLE
Add assembly attribute with name pattern for exporting

### ISSUE THREAD

**Issue by OmegaExtern** [3F/DllExport#12]
For example:
Say, assembly is marked: `[assembly: DllExportInclude("exportme_*")]`
It would export all static methods from assembly whose name matches the specified pattern (in this case `exportme_*`).

It would also be nice to have `DllExportExclude` attribute which could be applied to assembly, types, methods (etc.), which would work exactly the opposite of `DllExportInclude`.

**Comment by 3F** [comment]
> It would also be nice to have 

I don't think so... I mean the export-functions are not regular or common thing for something like this.

It's the same as to declare all protected methods as virtual and control all this via black/white list.

also will be a big surprise what's going on, for anyone else who will see source code in first time

and also important problem will be with the exported names by default:

your #10 related

How about:

```csharp
public static void Print(int a) { }
public static void Print(bool b) { }
```

What name the Dllexport should set as export-functions for first and second C# method by default ?

**Comment by OmegaExtern** [comment]
> What name the Dllexport should set as export-functions for first and second C# method by default ?

Their mangled names ;)

**Comment by 3F** [comment]
> Their mangled names ;)

:) good 

how about to help me first with here https://github.com/3F/Conari/issues/3 (Undecorate functions from C++ compilers)

I already have impl. of C linkage for __cdecl, __stdcall, __fastcall, __vectorcall

but as you should know the any overloading for this is not supported by their format.

so... C++ mangling for different compilers - is there any open source implementations of this ?

btw, in Visual Studio folder we can find console tool - undname (for Dllexport it probably a good way)

> It is already a surprise since metadata for DllExport is removed from compiled assembly...

it's not related for source code. For below, we can see export-func for dll:


```csharp
[DllExport]
public static bool ess(string data, string filter, bool ignoreCase)
{
    
}
```

for C++

```cpp
REGXWILD_API bool ess(const tstring* data, const tstring* filter, bool ignoreCase);
```

without DllExport or REGXWILD_API above it can be problem for some cases

**Comment by OmegaExtern** [comment]
> it's not related for source code.

See my previous comment (made 2nd edit).

> without DllExport or REGXWILD_API above it can be problem for some cases

Again, then it is not your nor mine problem, if somebody goes up to strange code-base and he/she looks at source code for the first time, there will be already plenty of surprises, and **especially** for new programmers, who are not even aware of the power of attributes and MSIL injection... (So, just you know, I disagree with your point there.)

P.S. Some obfuscators already provide such functionality (include/exclude and pattern matching) through (assembly) attributes.

> C++ mangling for different compilers - is there any open source implementations of this ?

I know of [demangler](https://demangler.com/), but I have never looked into name mangling much, in short, it enables the concept of polymorphism, etc.
There might be something about this in Roslyn, not sure.
Anyway, what is stopping you from doing own/DllExport (.NET) name (de)mangler?

**Comment by 3F** [comment]
I'm still not sure that's useful way, and probably I see more problems than good.

but ok, I marked it as a [suggestion], so it can be considered later.


> what is stopping you from doing own/DllExport (.NET) name (de)mangler?

my conscience :)

*for Conari, it other history of course.*
